tests/integration_tests/functional/test_config_space.py

```diff
@@ -0,0 +1,351 @@
+    # client socket connection, in case the server had not started yet.
+    _send_data_g2h(
+        ssh_control_connection,
+        host_ip,
+        host_port,
+        iterations,
+        data,
+        retries=5
+    )
+
+    # Wait for host server to receive the data sent by the guest client.
+    server_thread.join()
+
+
+def _change_guest_if_mac(ssh_connection, guest_if_mac, guest_if_name):
+    cmd = "ip link set dev {} address ".format(guest_if_name) + guest_if_mac
+    # The connection will be down, because changing the mac will issue down/up
+    # on the interface.
+    ssh_connection.execute_command(cmd)
+
+
+def _get_net_mem_addr_base_on_aarch64(ssh_connection, if_name):
+    """Get the net device memory start address."""
+    sys_virtio_mmio_cmdline = "/sys/devices/platform"
+    cmd = "ls {} | grep .virtio_mmio".format(sys_virtio_mmio_cmdline)
+    rc, stdout, _ = ssh_connection.execute_command(cmd)
+    assert rc == 0
+
+    virtio_devs = stdout.read().split()
+    devs_addr = list(map(lambda dev: dev.split(".")[0], virtio_devs))
+
+    cmd = "ls {}/{}/virtio{}/net"
+    # Device start addresses lack the hex prefix and are not interpreted
+    # accordingly when parsed inside `change_config_space.c`.
+    hex_prefix = '0x'
+    for (idx, dev) in enumerate(virtio_devs):
+        _, guest_if_name, _ = ssh_connection.execute_command(
+            cmd.format(sys_virtio_mmio_cmdline, dev, idx)
+        )
+        if guest_if_name.read().strip() == if_name:
+            return hex_prefix + devs_addr[int(idx)]
+    return None
+
+
+def _get_net_mem_addr_base_on_x86_64(ssh_connection, if_name):
+    """Get the net device memory start address."""
+    sys_virtio_mmio_cmdline = "/sys/devices/virtio-mmio-cmdline/"
+    cmd = "ls {} | grep virtio-mmio. | sed 's/virtio-mmio.//'"
+    exit_code, stdout, _ = ssh_connection.execute_command(
+        cmd.format(sys_virtio_mmio_cmdline)
+    )
+    assert exit_code == 0
+    virtio_devs_idx = stdout.read().split()
+
+    cmd = "cat /proc/cmdline"
+    exit_code, cmd_line, _ = ssh_connection.execute_command(cmd)
+    assert exit_code == 0
+    pattern_dev = re.compile("(virtio_mmio.device=4K@0x[0-9a-f]+:[0-9]+)+")
+    pattern_addr = re.compile("virtio_mmio.device=4K@(0x[0-9a-f]+):[0-9]+")
+    devs_addr = []
+    for dev in re.findall(pattern_dev, cmd_line.read()):
+        matched_addr = pattern_addr.search(dev)
+        # The 1st group which matches this pattern
+        # is the device start address. `0` group is
+        # full match.
+        addr = matched_addr.group(1)
+        devs_addr.append(addr)
+
+    cmd = "ls {}/virtio-mmio.{}/virtio{}/net"
+    for idx in virtio_devs_idx:
+        _, guest_if_name, _ = ssh_connection.execute_command(
+            cmd.format(sys_virtio_mmio_cmdline, idx, idx)
+        )
+        if guest_if_name.read().strip() == if_name:
+            return devs_addr[int(idx)]
+    return None
+
+
+def _get_net_if_name(ssh_connection, guest_ip):
```

Comment:
This looks like it belongs in `host_tools/network.py`

Reply:
Done.